File: TBOS_Backend/apps/reviews/services/review_service.py
from django.db import models, transaction

from apps.courses.models import Course
from apps.reviews.models import Review, ReviewResponse


class ReviewService:

    @staticmethod
    @transaction.atomic
    def create_review(student, course, rating, review_text):
        review = Review.objects.create(
            student=student,
            course=course,
            rating=rating,
            review_text=review_text,
        )
        ReviewService._refresh_course_rating(course)
        return review

    @staticmethod
    @transaction.atomic
    def update_review(review, rating=None, review_text=None):
        if rating is not None:
            review.rating = rating
        if review_text is not None:
            review.review_text = review_text
        review.save(update_fields=["rating", "review_text", "updated_at"])
        ReviewService._refresh_course_rating(review.course)
        return review

    @staticmethod
    @transaction.atomic
    def delete_review(review):
        course = review.course
        review.delete()
        ReviewService._refresh_course_rating(course)

    @staticmethod
    @transaction.atomic
    def add_instructor_response(review, instructor, response_text):
        response, created = ReviewResponse.objects.update_or_create(
            review=review,
            defaults={
                "instructor": instructor,
                "response_text": response_text,
            },
        )
        return response

    @staticmethod
    def calculate_course_rating(course):
        agg = Review.objects.filter(
            course=course, status=Review.Status.PUBLISHED
        ).aggregate(
            avg=models.Avg("rating"),
            count=models.Count("id"),
        )
        return {
            "average_rating": round(agg["avg"] or 0, 2),
            "rating_count": agg["count"] or 0,
        }

    @staticmethod
    def get_course_reviews(course):
        return (
            Review.objects.filter(course=course, status=Review.Status.PUBLISHED)
            .select_related("student")
            .prefetch_related("instructor_response__instructor")
            .order_by("-created_at")
        )

    @staticmethod
    def _refresh_course_rating(course):
        stats = ReviewService.calculate_course_rating(course)
        course.average_rating = stats["average_rating"]
        course.rating_count = stats["rating_count"]
        course.save(update_fields=["average_rating", "rating_count"])

File: tbos-frontend/services/reviewService.ts
// ============================================================
// services/reviewService.ts
// ============================================================

import { api, publicApi } from "@/lib/axios";
import type { ApiSuccess, PaginatedResponse } from "@/types/api";

export interface Review {
  id: string;
  student: string;
  student_name: string;
  course: string;
  rating: number;
  review_text: string;
  status: "published" | "pending" | "rejected";
  instructor_response: {
    id: string;
    instructor_name: string;
    response_text: string;
    created_at: string;
  } | null;
  created_at: string;
  updated_at: string;
}

export interface ReviewCreatePayload {
  course_id: string;
  rating: number;
  review_text: string;
}

export interface ReviewUpdatePayload {
  rating?: number;
  review_text?: string;
}

const ReviewService = {
  /** Public: Get all published reviews for a course */
  async getCourseReviews(courseId: string): Promise<PaginatedResponse<Review>> {
    const { data } = await publicApi.get<ApiSuccess<PaginatedResponse<Review>>>(
      "/reviews/public/",
      { params: { course: courseId } }
    );
    return data.data;
  },

  /** Student: Get my reviews */
  async getMyReviews(): Promise<PaginatedResponse<Review>> {
    const { data } = await api.get<ApiSuccess<PaginatedResponse<Review>>>(
      "/reviews/student/"
    );
    return data.data;
  },

  /** Student: Create review */
  async createReview(payload: ReviewCreatePayload): Promise<Review> {
    const { data } = await api.post<ApiSuccess<Review>>(
      "/reviews/student/",
      payload
    );
    return data.data;
  },

  /** Student: Update review */
  async updateReview(id: string, payload: ReviewUpdatePayload): Promise<Review> {
    const { data } = await api.patch<ApiSuccess<Review>>(
      `/reviews/student/${id}/`,
      payload
    );
    return data.data;
  },

  /** Student: Delete review */
  async deleteReview(id: string): Promise<void> {
    await api.delete(`/reviews/student/${id}/`);
  },

  /** Instructor: List reviews for own courses */
  async listInstructorReviews(courseId?: string): Promise<PaginatedResponse<Review>> {
    const { data } = await api.get<ApiSuccess<PaginatedResponse<Review>>>(
      "/reviews/instructor/",
      { params: courseId ? { course: courseId } : {} }
    );
    return data.data;
  },

  /** Instructor: Respond to review */
  async respondToReview(
    reviewId: string,
    responseText: string
  ): Promise<Review> {
    const { data } = await api.post<ApiSuccess<Review>>(
      "/reviews/instructor/respond/",
      { review_id: reviewId, response_text: responseText }
    );
    return data.data;
  },

  /** Admin: List all reviews */
  async listAdminReviews(params?: {
    status?: string;
    course?: string;
  }): Promise<PaginatedResponse<Review>> {
    const { data } = await api.get<ApiSuccess<PaginatedResponse<Review>>>(
      "/reviews/admin/",
      { params }
    );
    return data.data;
  },

  /** Admin: Moderate review */
  async moderateReview(
    id: string,
    status: "published" | "rejected"
  ): Promise<Review> {
    const { data } = await api.put<ApiSuccess<Review>>(
      `/reviews/admin/${id}/moderate/`,
      { status }
    );
    return data.data;
  },
};

export default ReviewService;
